### DISCUSSION TITLE
not working proxyMap<K, V>() with proxyWithHistory

### DISCUSSION THREAD

**Discussion by Danue1** [pmndrs/valtio#728] [Q&A] [answered]
Full Example: https://codesandbox.io/s/valtio-forked-purfbz?file=/src/App.tsx

We can follow the steps below on Full Example.
- Click "add"
- Click "add"
- Click "undo" <<- boom with "Cannot destructure property 'name' of '(0 , _valtio.useSnapshot)(...)' as it is undefined."

We can look at the source code and understand why this happens.
"deepClone" function is because it does not restore the proxy state.

https://github.com/pmndrs/valtio/blob/f8a1bdcf03e2c6ec0482f6aef051400f7881c826/src/vanilla/utils/proxyWithHistory.ts#L7-L21

It seems that this may be the design intent.
But what if we need a Map<K, V>?
1. Use Record<K, V> instead of Map<K, V>.
2. Modify proxyWithHistory to support Map<K, V>, Set<K>.
3. (no idea)

I'd like to avoid 1 if possible.

**Comment by dai-shi** [comment, marked as answer]
I don't think proxyWithHistory with embedded proxy is ever supported.
So, maybe you just don't use proxyWithHistory and implement history capability somehow on your end. (I don't have any concrete ideas.)

By design, proxyWithHistory is not perfect :point_right: https://github.com/pmndrs/valtio/discussions/725#discussioncomment-5822747

**Reply by Danue1** [reply]
Hmm, It is possible to stack only patches using immer.js, but considering the convenience of valtio, it seems somewhat disappointing. [ref.](https://itnext.io/easy-undo-feature-with-react-and-immer-js-9b467c080257)

Is there a better way?
For example, in valtio, there is an idea to provide an API that diffs only patches.

**Reply by dai-shi** [reply]
Creating patches is out of the scope of Valtio. The simplicity of Valtio design is snapshots, which cares only the result.
Though, we have a project outside to explore applying _kind of_ patches. (I wouldn't call/consider them "patches")
- https://github.com/jotaijs/jotai-valtio
- https://github.com/dai-shi/valtio-yjs

proxyWithHistory is designed around snapshots and that's why I suggest people to create a third-party package for more capability.

**Reply by Danue1** [reply]
Thank you for introducing the project, and understanding the secret of valtio's simplicity!
